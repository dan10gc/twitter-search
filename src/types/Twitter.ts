export interface HashtagModel {
    text: string;
    indices: Array<number>;
}

export interface MetaDataModel {
    completed_in: number;
    count: number;
    max_id: number;
    max_id_str: string;
    // is not returned when user reaches end of list
    next_results?: string;
    query: string
    since_id: number
    since_id_str: string
}