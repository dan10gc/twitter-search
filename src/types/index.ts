export * from './Status'
export * from './Twitter';