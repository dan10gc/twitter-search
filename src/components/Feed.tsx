import React from 'react';

import { MetaDataModel, Status } from '../types';
import { Spinner } from './Spinner';
import Tweet from './Tweet';

interface Props {
	tweets: Array<any>;
	status: Status;
	onLoadMore: () => void;
	paginationStatus: Status;
	metaData: MetaDataModel | null;
}

export const Feed = ({ tweets, status, onLoadMore, paginationStatus, metaData }: Props) => {
	return (
		<div className="box px-0 mb-5 ">
			{Status.PENDING === status && <Spinner />}

			{tweets.map((tweet, index) => {
				return <Tweet {...{ tweet, index }} key={tweet.user?.screen_name + index.toString()} />;
			})}
			{tweets.length > 0 && (
				<button
					className={`button is-ghost mt-4 is-fullwidth ${
						Status.PENDING === paginationStatus && `is-loading`
					}`}
					onClick={onLoadMore}
				>
					{/* @ts-ignore */}
					{!metaData?.next_results ? `Back to top` : `Load more`}
				</button>
			)}
		</div>
	);
};
