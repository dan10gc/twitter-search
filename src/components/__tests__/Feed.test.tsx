import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Feed } from '../Feed';
import { MetaDataModel, Status } from '../../types';

const onLoadMoreMock = jest.fn();

// let emptyTweetsArr: Array<any> = [];
const mockTweets: Array<any> = [
	{
		entities: { hashtags: [] },
		text:
			'Eight years ago today, Lionel Messi scored four goals for Barcelona against Osasuna ⚽️⚽️⚽️⚽️ The way he sat down thttps://t.co/z81vXrJZnz',
		user: {
			screen_name: 'goal',
			profile_image_url: 'http://pbs.twimg.com/profile_images/1306259078902165515/NFreUk7w_normal.jpg',
		},
	},
	{
		entities: { hashtags: [{ text: 'Messi', indices: [75, 101] }] },
		text: `All 👀's on #Messi https://t.co/akRMPS8eaw`,
		user: {
			screen_name: 'FCBarcelona',
			profile_image_url: 'http://pbs.twimg.com/profile_images/1333096463916797954/7bzarkH2_normal.jpg',
		},
	},
	{
		entities: { hashtags: [{ text: 'Messi', indices: [75, 101] }] },
		text: '💪 #Messi is back! 🐐 https://t.co/QyEDYPTNp7',
		user: {
			screen_name: 'FCBarcelona',
			profile_image_url: 'http://pbs.twimg.com/profile_images/1333096463916797954/7bzarkH2_normal.jpg',
		},
	},
	{
		entities: { hashtags: [] },
		text: `Kauan Basile is eight years old. He's just signed an apparel deal with Nike (which is younger than Neymar and Messi… https://t.co/uknJXTrtHr`,
		user: {
			screen_name: 'brfootball',
			profile_image_url: 'http://pbs.twimg.com/profile_images/1265929170913296385/I54xoXdG_normal.jpg',
		},
	},
	{
		entities: { hashtags: [] },
		text: `📈 Top rated players in Europe's top 5 leagues in 2021 (3+ apps): 1️⃣ L. Messi - 8.80 2️⃣ H. Kane - 8.46 3️⃣ J. San… https://t.co/nbkMTU0ipT`,
		user: {
			screen_name: 'WhoScored',
			profile_image_url: 'http://pbs.twimg.com/profile_images/1020650459349110784/EFuffquU_normal.jpg',
		},
	},
];

const mockMetaData: MetaDataModel = {
	completed_in: 0.025,
	count: 5,
	max_id: 0,
	max_id_str: '0',
	next_results: '?max_id=1354055253520482310&q=messi&count=5&include_entities=1&result_type=popular',
	query: 'messi',
	since_id: 0,
	since_id_str: '0',
};

describe('Feed Component', () => {
	test('renders idleMessage', () => {
		const { container, getByText } = render(
			<Feed
				tweets={[]}
				status={Status.IDLE}
				onLoadMore={onLoadMoreMock}
				paginationStatus={Status.IDLE}
				metaData={null}
			/>
		);
		expect(getByText('Start your search')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
	test('renders noResultsMessage', () => {
		const { container, getByText } = render(
			<Feed
				tweets={[]}
				status={Status.RESOLVED}
				onLoadMore={onLoadMoreMock}
				paginationStatus={Status.IDLE}
				metaData={null}
			/>
		);
		expect(getByText('No results')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
	test('renders 5 tweets and load more button', () => {
		const { container, getByText } = render(
			<Feed
				tweets={mockTweets}
				status={Status.RESOLVED}
				onLoadMore={onLoadMoreMock}
				paginationStatus={Status.IDLE}
				metaData={mockMetaData}
			/>
		);
		expect(getByText('Load more')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
});
