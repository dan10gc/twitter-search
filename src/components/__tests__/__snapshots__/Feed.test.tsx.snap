// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed Component renders 5 tweets and load more button 1`] = `
<div>
  <div
    class="box px-0 mb-5 "
  >
    <div
      class="box mb-0 px-mobile is-shadowless is-radiusless false"
    >
      <article
        class="media"
      >
        <div
          class="media-left"
        >
          <figure
            class="image is-48x48"
          >
            <img
              alt="user profile"
              class="is-rounded"
              src="http://pbs.twimg.com/profile_images/1306259078902165515/NFreUk7w_normal.jpg"
            />
          </figure>
        </div>
        <div
          class="media-content"
        >
          <div
            class="content"
          >
            <h5
              class="title is-5 mb-2"
            >
              <strong>
                @
                goal
              </strong>
            </h5>
            <p
              class="mb-1 has-text-weight-medium tweet-text"
            >
              Eight years ago today, Lionel Messi scored four goals for Barcelona against Osasuna ‚öΩÔ∏è‚öΩÔ∏è‚öΩÔ∏è‚öΩÔ∏è The way he sat down t
            </p>
            <a
              href="https://t.co/z81vXrJZnz"
              rel="noreferrer"
              target="_blank"
            >
              https://t.co/z81vXrJZnz
            </a>
          </div>
          <div
            class="is-flex-direction-row flex-wrap is-justify-content-start"
          />
        </div>
      </article>
    </div>
    <div
      class="box mb-0 px-mobile is-shadowless is-radiusless has-background-white-bis"
    >
      <article
        class="media"
      >
        <div
          class="media-left"
        >
          <figure
            class="image is-48x48"
          >
            <img
              alt="user profile"
              class="is-rounded"
              src="http://pbs.twimg.com/profile_images/1333096463916797954/7bzarkH2_normal.jpg"
            />
          </figure>
        </div>
        <div
          class="media-content"
        >
          <div
            class="content"
          >
            <h5
              class="title is-5 mb-2"
            >
              <strong>
                @
                FCBarcelona
              </strong>
            </h5>
            <p
              class="mb-1 has-text-weight-medium tweet-text"
            >
              All üëÄ's on #Messi 
            </p>
            <a
              href="https://t.co/akRMPS8eaw"
              rel="noreferrer"
              target="_blank"
            >
              https://t.co/akRMPS8eaw
            </a>
          </div>
          <div
            class="is-flex-direction-row flex-wrap is-justify-content-start"
          >
            <div
              class="badge is-clickable"
            >
              <span>
                 #
                Messi
              </span>
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="box mb-0 px-mobile is-shadowless is-radiusless false"
    >
      <article
        class="media"
      >
        <div
          class="media-left"
        >
          <figure
            class="image is-48x48"
          >
            <img
              alt="user profile"
              class="is-rounded"
              src="http://pbs.twimg.com/profile_images/1333096463916797954/7bzarkH2_normal.jpg"
            />
          </figure>
        </div>
        <div
          class="media-content"
        >
          <div
            class="content"
          >
            <h5
              class="title is-5 mb-2"
            >
              <strong>
                @
                FCBarcelona
              </strong>
            </h5>
            <p
              class="mb-1 has-text-weight-medium tweet-text"
            >
              üí™ #Messi is back! üêê 
            </p>
            <a
              href="https://t.co/QyEDYPTNp7"
              rel="noreferrer"
              target="_blank"
            >
              https://t.co/QyEDYPTNp7
            </a>
          </div>
          <div
            class="is-flex-direction-row flex-wrap is-justify-content-start"
          >
            <div
              class="badge is-clickable"
            >
              <span>
                 #
                Messi
              </span>
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="box mb-0 px-mobile is-shadowless is-radiusless has-background-white-bis"
    >
      <article
        class="media"
      >
        <div
          class="media-left"
        >
          <figure
            class="image is-48x48"
          >
            <img
              alt="user profile"
              class="is-rounded"
              src="http://pbs.twimg.com/profile_images/1265929170913296385/I54xoXdG_normal.jpg"
            />
          </figure>
        </div>
        <div
          class="media-content"
        >
          <div
            class="content"
          >
            <h5
              class="title is-5 mb-2"
            >
              <strong>
                @
                brfootball
              </strong>
            </h5>
            <p
              class="mb-1 has-text-weight-medium tweet-text"
            >
              Kauan Basile is eight years old. He's just signed an apparel deal with Nike (which is younger than Neymar and Messi‚Ä¶ 
            </p>
            <a
              href="https://t.co/uknJXTrtHr"
              rel="noreferrer"
              target="_blank"
            >
              https://t.co/uknJXTrtHr
            </a>
          </div>
          <div
            class="is-flex-direction-row flex-wrap is-justify-content-start"
          />
        </div>
      </article>
    </div>
    <div
      class="box mb-0 px-mobile is-shadowless is-radiusless false"
    >
      <article
        class="media"
      >
        <div
          class="media-left"
        >
          <figure
            class="image is-48x48"
          >
            <img
              alt="user profile"
              class="is-rounded"
              src="http://pbs.twimg.com/profile_images/1020650459349110784/EFuffquU_normal.jpg"
            />
          </figure>
        </div>
        <div
          class="media-content"
        >
          <div
            class="content"
          >
            <h5
              class="title is-5 mb-2"
            >
              <strong>
                @
                WhoScored
              </strong>
            </h5>
            <p
              class="mb-1 has-text-weight-medium tweet-text"
            >
              üìà Top rated players in Europe's top 5 leagues in 2021 (3+ apps): 1Ô∏è‚É£ L. Messi - 8.80 2Ô∏è‚É£ H. Kane - 8.46 3Ô∏è‚É£ J. San‚Ä¶ 
            </p>
            <a
              href="https://t.co/nbkMTU0ipT"
              rel="noreferrer"
              target="_blank"
            >
              https://t.co/nbkMTU0ipT
            </a>
          </div>
          <div
            class="is-flex-direction-row flex-wrap is-justify-content-start"
          />
        </div>
      </article>
    </div>
    <button
      class="button is-ghost mt-4 is-fullwidth false"
    >
      Load more
    </button>
  </div>
</div>
`;

exports[`Feed Component renders idleMessage 1`] = `
<div>
  <div
    class="box px-0 mb-5 "
  >
    <div
      class="content py-4"
    >
      <figure
        class="image is-400x400"
      >
        <img
          alt="analytics"
          src="https://ik.imagekit.io/tb8dhpn3xiv/graphs_statistics_outline_II_q-M4DXaqR.png"
        />
      </figure>
      <div
        class="px-5"
      >
        <h1>
          Start your search
        </h1>
        <p>
          Search Twitter for keywords and monitor trends and hashtags in your industry.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feed Component renders noResultsMessage 1`] = `
<div>
  <div
    class="box px-0 mb-5 "
  >
    <div
      class="content py-4"
    >
      <figure
        class="image is-400x400"
      >
        <img
          alt="search"
          src="https://ik.imagekit.io/tb8dhpn3xiv/search_outline_II_gW4C46dD5.png"
        />
      </figure>
      <div
        class="px-5"
      >
        <h1>
          No results
        </h1>
        <p>
          The term you entered did not bring up any results, you may have mistyped your term.
        </p>
      </div>
    </div>
  </div>
</div>
`;
